web/src/ui.rs

```diff
@@ -116,6 +120,91 @@ impl FileDialogResult for WebFileDialogResult {
     fn refresh(&mut self) {}
 }
 
+pub struct WebDownloadFileData {
+    file_name: String,
+    contents: Vec<u8>,
+    downloaded: bool,
+}
+
+pub struct WebDownloadFileDialogResult {
+    data: Mutex<WebDownloadFileData>,
+}
+
+impl WebDownloadFileDialogResult {
+    fn new(file_name: String) -> WebDownloadFileDialogResult {
+        Self {
+            data: Mutex::new(WebDownloadFileData {
+                file_name,
+                contents: Vec::new(),
+                downloaded: false,
+            }),
+        }
+    }
+
+    fn data(&self) -> MutexGuard<'_, WebDownloadFileData> {
+        self.data.lock().expect("non-poisoned data")
+    }
+
+    fn download(mut data: MutexGuard<WebDownloadFileData>) -> Option<()> {
+        let array = Uint8Array::from(&data.contents[..]);
+        let blob = Blob::new_with_u8_array_sequence(&array).ok()?;
+        let window = web_sys::window()?;
+        let document = window.document()?;
+        let a = document.create_element("a").ok()?;
+
+        let url = JsUrl::create_object_url_with_blob(&blob).ok()?;
+        a.set_attribute("href", &url).ok()?;
+        a.set_attribute("download", &data.file_name).ok()?;
+        a.dyn_into::<HtmlElement>().ok()?.click();
+        data.downloaded = true;
+        Some(())
+    }
+}
+
+impl FileDialogResult for WebDownloadFileDialogResult {
+    fn is_cancelled(&self) -> bool {
+        self.data().downloaded
```

Comment:
I am not sure if that is correct

Reply:
That was not correct, it was my first assumption about the code